lib/lf/uscalfe/lagr_fe.h

```diff
@@ -1155,6 +1155,500 @@ const Eigen::MatrixXi FeLagrangeO2Quad<SCALAR>::ksegment_to_quad_mapping_ =
+
+    // Evaluate basis functions using the tensor product structure
+    for (int i = 0; i < 16; ++i) {
+      result.row(i) = (segment_x0_eval.row(ksegment_to_quad_mapping_(i, 0)) *
+                       segment_x1_eval.row(ksegment_to_quad_mapping_(i, 1)))
+                          .matrix();
+    }
+    return result;
+  }
+
+  [[nodiscard]] Eigen::Matrix<SCALAR, Eigen::Dynamic, Eigen::Dynamic>
+  GradientsReferenceShapeFunctions(
+      const Eigen::MatrixXd& refcoords) const override {
+    LF_ASSERT_MSG(refcoords.rows() == 2,
+                  "Reference coordinates must be 2-vectors");
+
+    const size_type n_pts(refcoords.cols());
+    Eigen::Matrix<SCALAR, Eigen::Dynamic, Eigen::Dynamic> result(16, 2 * n_pts);
+
+    // reshape into a 32xn_pts matrix.
+    Eigen::Map<Eigen::Array<SCALAR, Eigen::Dynamic, Eigen::Dynamic>,
+               Eigen::AutoAlign>
+        temp(&result(0, 0), 32, n_pts);
+
+    // evaluate "segment" shape functions (b^j and b^l)
+    Eigen::Array<SCALAR, Eigen::Dynamic, Eigen::Dynamic> segment_x0_eval =
+        (krsf_segment_.EvalReferenceShapeFunctions(refcoords.row(0))).array();
+    Eigen::Array<SCALAR, Eigen::Dynamic, Eigen::Dynamic> segment_x1_eval =
+        (krsf_segment_.EvalReferenceShapeFunctions(refcoords.row(1))).array();
+
+    // evaluate derivatives of "segment" shape functions (b^j and
+    // b^l)
+    Eigen::Array<SCALAR, Eigen::Dynamic, Eigen::Dynamic> segment_x0_grad =
+        (krsf_segment_.GradientsReferenceShapeFunctions(refcoords.row(0)))
+            .array();
+    Eigen::Array<SCALAR, Eigen::Dynamic, Eigen::Dynamic> segment_x1_grad =
+        (krsf_segment_.GradientsReferenceShapeFunctions(refcoords.row(1)))
+            .array();
+
+    // evaluate gradients  using the product rule and
+    // the tensor product structure of the basis functions.
+    for (int i = 0; i < 16; ++i) {
+      // d/dx
+      temp.row(i) = (segment_x0_grad.row(ksegment_to_quad_mapping_(i, 0)) *
+                     segment_x1_eval.row(ksegment_to_quad_mapping_(i, 1)))
+                        .matrix();
+      // d/dy
+      temp.row(i + 16) = (segment_x1_grad.row(ksegment_to_quad_mapping_(i, 1)) *
+                          segment_x0_eval.row(ksegment_to_quad_mapping_(i, 0)))
+                             .matrix();
+    }
+
+    return result;
+  }
+
+  [[nodiscard]] Eigen::MatrixXd EvaluationNodes() const override {
+    Eigen::MatrixXd nodes(2, 16);
+
+    Eigen::MatrixXd vertices(2, 4);
+    Eigen::MatrixXd midpoints(2, 8);
+    Eigen::MatrixXd interior(2, 4);
+
+    // clang-format off
+    vertices << 0.0, 1.0, 1.0, 0.0,
+               0.0, 0.0, 1.0, 1.0;
+    midpoints << 1.0, 2.0, 3.0, 3.0, 2.0, 1.0, 0.0, 0.0,
+                 0.0, 0.0, 1.0, 2.0, 3.0, 3.0, 2.0, 1.0;
+    interior << 1.0, 2.0, 2.0, 1.0,
+                1.0, 1.0, 2.0, 2.0;
+    // clang-format on
+    midpoints *= 1.0 / 3.0;
+    interior *= 1.0 / 3.0;
+
```

Comment:
What about using `constexpr` to "encourage" the compiler to do all evaluations at compile time. Should work at least with fixed-size matrices and for `vertices`, `midpoints`, and `interior`.

Reply:
I couldn't find a lot of information about this, but as far as I know, `Eigen` matrices do not have  `constexpr` constructors/methods (?).

 Another possibility to improve the performance of this function could be to store the evaluation nodes as another static member field. However this could make the implementation of this function unnecessarily fragmented.